File: 2284_SenderWithLargestWordCount/2284_SenderWithLargestWordCount.js
/**
 * @param {string[]} messages
 * @param {string[]} senders
 * @return {string}
 */
var largestWordCount = function (messages, senders) {
  let cache = {};
  for (let i = 0; i < messages.length; i++) {
    const count = messages[i].split(' ').length;
    cache[senders[i]] = (cache[senders[i]] || 0) + count;
  }
  let maxCount = 0;
  let resultSender = '';

  // find the sender with the largest word count, with lexicographically largest name in case of a tie
  for (let sender in cache) {
    let count = cache[sender];
    if (count > maxCount || (count === maxCount && sender > resultSender)) {
      maxCount = count;
      resultSender = sender;
    }
  }
  return resultSender;
};

File: 2284_SenderWithLargestWordCount/2284_SenderWithLargestWordCount.py
class Solution:
    def largestWordCount(self, messages: List[str], senders: List[str]) -> str:
        word_count = {}
        #word_count = defaultdict(int)
        # this provides a default value for nonexistent keys to avoid KeyERror. 
        #when you access a key that does not exist tin the dictionary, it will automatically intiilize with the default value of int()
        for i in range(len(messages)):
            count = len(messages[i].split())
            word_count[senders[i]] = word_count.get(senders[i], 0) + count
        
        maxWordCount = 0
        maxSender = ""
        
        for sender, count in word_count.items():
            if count > maxWordCount or (count == maxWordCount and sender > maxSender):
                maxWordCount = count
                maxSender = sender
        return maxSender